m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -0,0 +1,781 @@
+                break;
+            case GAUGE:
+                GaugeValue gaugeValue = resourcePool.getGaugeValue();
+                gaugeValue.setDValue(dValue);
+                metricCopy.getMetricValue().setGauge(gaugeValue);
+                break;
+            case TIMER:
+                TimerValue timerValue = resourcePool.getTimerValue();
+                timerValue.setI64Value(iValue);
+                metricCopy.getMetricValue().setTimer(timerValue);
+                break;
+            default:
+                throw new IllegalArgumentException("Unsupported metric type: " + metricType);
+        }
+
+        try {
+            metricQueue.put(resourcePool.getSizedMetric().setMetric(metricCopy).setSize(size));
+        } catch (InterruptedException e) {
+            LOG.warn(String.format("Interrupted while putting %s: %s",
+                metricType,
+                metricCopy.getName()));
+        }
+    }
+
+    private String valueBucketString(double bucketBound) {
+        if (bucketBound == Double.MAX_VALUE) {
+            return "infinity";
+        }
+
+        if (bucketBound == -Double.MAX_VALUE) {
+            return "-infinity";
+        }
+
+        return String.format(bucketValFmt, bucketBound);
+    }
+
+    private String durationBucketString(Duration bucketBound) {
+        if (Duration.MAX_VALUE.equals(bucketBound)) {
+            return "infinity";
+        }
+
+        if (Duration.MIN_VALUE.equals(bucketBound)) {
+            return "-infinity";
+        }
+
+        return bucketBound.toString();
+    }
+
+    private class Processor implements Runnable {
+        @Override
+        public void run() {
+            try {
+                List<Metric> metrics = new ArrayList<>(freeBytes / 10);
+                int bytes = 0;
+
+                while (!metricQueue.isEmpty()) {
+                    SizedMetric sizedMetric = metricQueue.poll();
+                    Metric metric = sizedMetric.getMetric();
+                    int size = sizedMetric.getSize();
+
+                    resourcePool.releaseSizedMetric(sizedMetric);
+
+                    if (metric == null) {
+                        // Explicit flush requested
+                        if (metrics.size() > 0) {
+                            metrics.clear();
+                            bytes = 0;
+                        }
+
+                        continue;
+                    }
+
+                    if (bytes + size > freeBytes) {
+                        metrics = flush(metrics);
+                        bytes = 0;
+                    }
+
+                    metrics.add(metric);
+                    bytes += size;
+                }
+
+                if (metrics.size() > 0) {
+                    // Final flush
+                    flush(metrics);
+                }
+            } catch (TException e) {
+                throw new RuntimeException("Failed to emit metrics", e);
```

Comment:
A single exception shouldn't cause the processor to die, could you add a configurable retry?